File: Java/SlidingWindowAndTwoPointers/424-longest-repeating-character-replacement/longest-repeating-character-replacement.java
class Solution {
    static {
    for(int i = 0; i < 500; i++) {
        characterReplacement("", 1);
    }
    }
    public static int characterReplacement(String s, int k) {
        int n=s.length();
        int l=0, r=0, maxlen=0, maxf=0;
        int hash[] = new int[26];
        while(r<n){
            hash[s.charAt(r) - 'A']++;
            maxf = Math.max(maxf, hash[s.charAt(r) - 'A']);
            if((r-l+1) - maxf >k){
                hash[s.charAt(l) - 'A']--;
                l++;
            }
            maxlen = Math.max(maxlen, r-l+1);
            r++;
        }
        return maxlen;
    }
}

File: Python/SlidingWindowAndTwoPointers/Medium/424-longest-repeating-character-replacement/longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        l=0
        r=0
        n=len(s)
        
        maxlen=0
        mapp : Dict[str, int]= {}
        max_freq = 0
        while r < n:
            mapp[s[r]] = mapp.get(s[r],0) + 1
            max_freq = max(max_freq, mapp[s[r]])

            if (r - l + 1) - max_freq  > k:
                # while len(mapp) > k:
                mapp[s[l]]  -= 1
                if mapp[s[l]] == 0:
                    del mapp[s[l]]
                l+=1
            
            
            maxlen = max(maxlen, r-l+1)

            r+=1
            
        return maxlen